### PR TITLE
Make scalar coupleable return the scalar variable order

### LINKED ISSUE #6376: Need ScalarCoupleable to provide scalar variable order
As title. I will send a small PR for this.

### PR DESCRIPTION
**Pull request by YaqiWang**
Close  #6376.

BTW, a new accessor `MooseVariableScalar::getOrder()` is added in line of `MooseVariable`. It is actually duplicating `MooseVariableBase::order()`. But I am not removing the later because I am afraid that it could have impact on the framework or modules or applications which could be using it.

### CODE REVIEW

**Comment by andrsd** [comment]
Why can't you use `MooseVariableBase::order()`? It should give the correct variable order back.

**Comment by YaqiWang** [comment]
Yes it is, but I want this more strait-forward, otherwise we will need to get the `MooseVariableScalar` first then get the order. It is also consistent with `MooseVariable`.

**Comment by andrsd** [comment]
That's not my point. I am saying there is no need for additional `getOrder()`, because `MooseVariableBase::order()` already gives back the correct number. I am fine with enhancing `ScalarCoupleable` API.

File: framework/src/base/ScalarCoupleable.C [hunk 1]
```diff
@@ -84,6 +84,15 @@ ScalarCoupleable::coupledScalar(const std::string & var_name, unsigned int comp)
   return getScalarVar(var_name, comp)->number();
 }
 
+Order
+ScalarCoupleable::coupledScalarOrder(const std::string & var_name, unsigned int comp)
+{
+  if (!isCoupledScalar(var_name, comp))
+    return FIRST;
```
Comment (andrsd) [inline framework/src/base/ScalarCoupleable.C]: No!

**Comment by YaqiWang** [comment]
Actually I think we should remove `MooseVariableBase::order`.

File: framework/src/base/ScalarCoupleable.C (hunk 1 above)
Comment (YaqiWang) [inline framework/src/base/ScalarCoupleable.C]: Please! tell me why...

**Comment by andrsd** [comment]
> Actually I think we should remove MooseVariableBase::order.

Why?

**Comment by YaqiWang** [comment]
@friedmud added `getOrder()` in `MooseVariable`. I think this is the more strait-forward way. Variable is added by MOOSE and MOOSE knows its order. No need to dig into libMesh interface to get the order back.

**Comment by andrsd** [comment]
This needs to get straighten out, then. We should not need 2 functions doing the same thing...

File: framework/src/base/ScalarCoupleable.C (hunk 1 above)
Comment (andrsd) [inline framework/src/base/ScalarCoupleable.C]: If the variable is not coupled, then it is not of first order.

Comment (YaqiWang) [inline framework/src/base/ScalarCoupleable.C]: Actually it is with the default coupling unless I understand default coupling wrongly.

Comment (andrsd) [inline framework/src/base/ScalarCoupleable.C]: But default coupling in not always on, see [addCoupledVar(name, doc_string)](https://github.com/idaholab/moose/blob/devel/framework/include/utils/InputParameters.h#L267)

Comment (YaqiWang) [inline framework/src/base/ScalarCoupleable.C]: But that line is only for checking default coupling...

Comment (andrsd) [inline framework/src/base/ScalarCoupleable.C]: If I set an optionally coupled variable in `validParams` and I do not couple this var in the input file and call `coupledScalarOrder(that_var)` in c-tor initializer, I am gonna get FIRST back, which is IMO wrong, there should be an error message.

Comment (YaqiWang) [inline framework/src/base/ScalarCoupleable.C]: That is what `test\src\auxkernels\CoupledScalarAux.C` is doing. After looking into `ScalarCoupleable::getDefaultValue()`, the order here should be `FEProblem::getMaxScalarOrder()`.

**Comment by YaqiWang** [comment]
That can be handled in a separate issue. I mentioned this PR will not be addressing it.

File: framework/src/base/ScalarCoupleable.C (hunk 1 above)
Comment (andrsd) [inline framework/src/base/ScalarCoupleable.C]: On Tue, Feb 23, 2016 at 2:40 PM, Yaqi <notifications@github.com> wrote:

> In framework/src/base/ScalarCoupleable.C
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_idaholab_moose_pull_6377-23discussion-5Fr53854823&d=BQMCaQ&c=54IZrppPQZKX9mLzcGdPfFD1hxrcB__aEkJFOKJFd00&r=3VAyeCMeLVVC6uUgEL2ffJ0a2Q62YeuUUIYP8elb-II&m=JJwc9U8DYMU5GxHVonP1QVACE2KyH58R8-O7oobBVmU&s=vv7kXCF3dL6NGI4c06w4gUz4lyTM8Uy6yYFXmEOSQLk&e=>
> :
>
> > @@ -84,6 +84,15 @@ ScalarCoupleable::coupledScalar(const std::string & var_name, unsigned int comp)
> >    return getScalarVar(var_name, comp)->number();
> >  }
> >
> > +Order
> > +ScalarCoupleable::coupledScalarOrder(const std::string & var_name, unsigned int comp)
> > +{
> > +  if (!isCoupledScalar(var_name, comp))
> > +    return FIRST;
>
> That is what test\src\auxkernels\CoupledScalarAux.C is doing. After
> looking into ScalarCoupleable::getDefaultValue(), the order here should
> be FEProblem::getMaxScalarOrder().
>

​CoupledScalarAux​ _is_ using the default value version. I am objecting the
case when users do _NOT_ use the default value. In that case this method
will return an incorrect value.

**Comment by andrsd** [comment]
> That can be handled in a separate issue. I mentioned this PR will not be addressing it.

I still do not believe that `getOrder()` is the right method, `order()` does the right thing and lives in the base class. That information is stored in libMesh so there is not need to duplicate it in MOOSE. I am actually for removing `getOrder()` from `MooseVariable`. And still think that you can achieve this without adding the new `getOrder()`.

**Comment by YaqiWang** [comment]
The latest PR added a check in `ScalarCoupleable::isCoupledScalar()` for the case that a non-required variable is used but not provided in the input, as pointed by @andrsd .

**Comment by friedmud** [comment]
I don't understand the purpose of this PR at all.  `order()` already does the right thing and is properly implemented in terms of calling libMesh.  It exists because we shouldn't have to "dig" into libMesh for this information (just look at its implementation... We don't need to have copies of that everywhere).  It doesn't duplicate any data and is always correct.

Can you describe the purpose of this new function?  It duplicates an existing function, adds storage and can potentially be out of sync with what's stored in libMesh...

**Comment by YaqiWang** [comment]
First of all, I completely agree that duplication with `getOrder()` and `order()` is bad, which already caused confusion. You can search through the framework and both are used.

Is having a local copy of `fe_type` or `order` OK or not? I think in this case it is OK because
1. we will never have millions of variables;
2. neither moose and libMesh provides a mutator of them. So whether they in MOOSE and libMesh are in sync is framework developers task;
3. if somewhere this function was called millions of times during run-time, somewhere we would need to keep a copy to make their access more efficient. It would be preferred in `MooseVariable`.

In what type we should return `Order` or `unsigned int`? I think we should use `Order`. Then we will need to deal with the conversion between `Order` enum and `unsigned int`. We can have implicit conversion from `Order` to `unsigned int`. `unsigned int` to `Order` requires explicit conversion which is actually desired.

So in conclusion, I think that we should remove the current `order()` and move `getOrder()` in `MooseVariableBase`. (Both variable and scalar variable have FEType.) Then we rename `getOrder()` to `order()`.

If we decide to have this change, I will need to fix Yak first. I am not sure if other applications are impacted. Modules are not using this function. Fixing framework is simple. This is why I want to address it in another separate issue. I can move `getOrder()` in `MooseVariableBase` in this PR.

The purpose of this PR is to provide a simpler interface to get the coupled scalar variable order. Otherwise we will need to call `getScalarVar()->order()`, which also does not support default coupling.

**Comment by permcody** [comment]
OK, I'm going to step in here since it seems people are getting a little worked up. I think everyone here has valid points. Yaqi is right that we already have two methods in MOOSE for the nonlinear/auxvariables and we are using both (Yuck!). However, at least the two existing versions in MOOSE aren't creating extra storage. Both of them are returning the order from libMesh information just as different types. Yaqi, your method is adding mutable storage which as David points out, could be out of sync with libMesh so I'd rather not merge this as is. You're right however that we probably should clean up MOOSE. This isn't your responsibility to get rid of one of the two duplicate methods but we shouldn't make a bigger mess either. As you pointed out, `Order` can be implicitly converted to `unsigned int` so I don't think we need two.

So, It looks like the best path forward would be to move two a single interface that returns a type `Order` that does not create extra storage, _period_.

So, let's do this right the first time for `ScalarVariables`. Let's move the existing getMethod variable from `MooseVariable` to `MooseVariableBase` unless there's something that doesn't allow that to work.

We'll open a ticket to deprecate the existing API and move people away from it.
Is everyone on board?

**Comment by YaqiWang** [comment]
@permcody a small correction, actually `MooseVariable::getOrder()` does store `fetype` currently. The current `MooseVariableBase::order()` is pretty ugly. I am ok with the rest.

Do you think creating this new interface in `ScalarCoupleable` `coupledScalarOrder` a good idea for supporting default values? I do not need this feature very much but just thought it could be useful.

**Comment by permcody** [comment]
I don't really care if it's ugly as long as it's working and not violating any normal coding principle. Yes, I see that `fetype` is stored. That probably could be better.

I'm not sure about the default `Order`, @andrsd may have an opinion on this.

**Comment by YaqiWang** [comment]
I finalized our discussions into three cis. Cody, sorry for not exactly following your suggestion on `order()`. We have this `fetype` stored anyhow.

**Comment by YaqiWang** [comment]
oh, this will require a patch in Yak, but I think the patch can be done without an integration branch in MOOSE.

File: framework/include/base/MooseVariableBase.h [hunk 2]
```diff
@@ -59,6 +60,11 @@ class MooseVariableBase
   unsigned int number() const { return _var_num; }
 
   /**
+   * Get the type of finite element object
+   */
+  const FEType feType() const { return _fe_type; }
```
Comment (permcody) [inline framework/include/base/MooseVariableBase.h]: I think you want to return a const reference here.

File: framework/include/base/MooseVariableBase.h [hunk 3]
```diff
@@ -85,8 +91,9 @@ class MooseVariableBase
 
   /**
    * Get the order of this variable
+   * Note: Order enum can be implicitly converted to unsigned int.
    */
-  unsigned int order() const;
+  Order order() const;
```
Comment (permcody) [inline framework/include/base/MooseVariableBase.h]: Normally I don't like to just change public interfaces in MOOSE like this but it looks like nobody is using this except you  (Yak)?

File: framework/include/base/MooseVariableBase.h [hunk 4]
```diff
@@ -106,6 +113,8 @@ class MooseVariableBase
 protected:
   /// variable number (from libMesh)
   unsigned int _var_num;
+  /// The FEType associated with this variable
+  FEType _fe_type;
```
Comment (permcody) [inline framework/include/base/MooseVariableBase.h]: Likewise, we can probably just store a const reference to this as well. I realize that we had a copy of it in `MooseVariable` but we may as well fix it.

File: framework/include/base/MooseVariableBase.h (hunk 3 above)
Comment (YaqiWang) [inline framework/include/base/MooseVariableBase.h]: It seems so...

File: framework/include/base/MooseVariableBase.h (hunk 4 above)
Comment (YaqiWang) [inline framework/include/base/MooseVariableBase.h]: I am changing it. I like to make it as a reference of the data in libmesh.

**Comment by YaqiWang** [comment]
Something weird happening: using the FEType obtained from libMesh causing strange errors...

**Comment by YaqiWang** [comment]
Changing the header files cause the rebuild of entire moose... can ccache help on this?

**Comment by permcody** [comment]
No, if you change the header file, it really does cascade down into several files because of #include

**Comment by YaqiWang** [comment]
Seems libMesh reconstructs variable FEType somewhere, which prevent us from creating a reference of it right after variable is added. So we will just make a copy. I will also leave a FIXME in the code as a reminder that we indeed made a copy in `MooseVariableBase` and things could be out of sync.

File: framework/src/bcs/DGFunctionDiffusionDirichletBC.C [hunk 5]
```diff
@@ -42,7 +42,7 @@ DGFunctionDiffusionDirichletBC::DGFunctionDiffusionDirichletBC(const InputParame
 Real
 DGFunctionDiffusionDirichletBC::computeQpResidual()
 {
-  const unsigned int elem_b_order = static_cast<unsigned int> (_var.getOrder());
+  const unsigned int elem_b_order = _var.order();
```
Comment (permcody) [inline framework/src/bcs/DGFunctionDiffusionDirichletBC.C]: @jwpeterson - what are your thoughts on removing this explicit cast. `order()` returns an `Order`. We had an explicit cast but it appears @YaqiWang likes the implicit conversion. I think the explicit one is probably better, but I'm willing to listen to arguments.

Comment (YaqiWang) [inline framework/src/bcs/DGFunctionDiffusionDirichletBC.C]: My only argument is that if we do it here then we need to everywhere, such as
```
for (unsigned int i=0; i<_var.order(); ++i)
```
which seems painful and unnecessary.

File: framework/include/base/MooseVariableBase.h (hunk 4 above)
Comment (YaqiWang) [inline framework/include/base/MooseVariableBase.h]: This has to be a copy, see my comments bellow.

File: framework/include/base/MooseVariableBase.h (hunk 3 above)
Comment (YaqiWang) [inline framework/include/base/MooseVariableBase.h]: I have patched Yak to make it only use `order()`.

**Comment by permcody** [comment]
This looks good. Thanks for jumping through all of these hoops.
:+1:

**Comment by andrsd** [comment]
:+1:  on the changes. I did not have internet access yesterday, so I could not respond.
